src/state/apis/swappers/helpers/getInputOutputRatioFromQuote.ts

```diff
@@ -94,21 +79,49 @@ const _getReceiveSideAmountsCryptoBaseUnit = ({
   quote: TradeQuote
   swapperName: SwapperName
 }) => {
+  const firstStep = quote.steps[0]
   const lastStep = quote.steps[quote.steps.length - 1]
   const slippageDecimalPercentage =
     quote.recommendedSlippage ?? getDefaultSlippagePercentageForSwapper(swapperName)
+  const rate = quote.rate
 
   const buyAmountCryptoBaseUnit = bn(lastStep.buyAmountBeforeFeesCryptoBaseUnit)
   const slippageAmountCryptoBaseUnit = buyAmountCryptoBaseUnit.times(slippageDecimalPercentage)
-  const buySideNetworkFeeCryptoBaseUnit = bn(0) // TODO(woodenfurniture): handle osmo swapper crazy network fee logic here
-  const buySideProtocolFeeCryptoBaseUnit = bnOrZero(
-    lastStep.feeData.protocolFees[lastStep.buyAsset.assetId]?.amountCryptoBaseUnit,
-  )
+  const sellAssetProtocolFee = firstStep.feeData.protocolFees[firstStep.sellAsset.assetId]
+  const buyAssetProtocolFee = lastStep.feeData.protocolFees[lastStep.buyAsset.assetId]
+  const sellSideProtocolFeeCryptoBaseUnit = bnOrZero(sellAssetProtocolFee?.amountCryptoBaseUnit)
+  const sellSideProtocolFeeBuyAssetBaseUnit = bnOrZero(
+    convertPrecision({
+      value: sellSideProtocolFeeCryptoBaseUnit,
+      inputExponent: firstStep.sellAsset.precision,
+      outputExponent: lastStep.buyAsset.precision,
+    }),
+  ).times(rate)
+  // Network fee represented as protocol fee for Osmosis swaps
+  const buySideNetworkFeeCryptoBaseUnit =
+    swapperName === SwapperName.Osmosis
+      ? (() => {
+          const isAtomOsmo = firstStep.sellAsset.chainId === cosmosChainId
+
+          // Subtract ATOM fees converted to OSMO for ATOM -> OSMO
+          if (isAtomOsmo) {
+            const otherDenomFee = lastStep.feeData.protocolFees[firstStep.sellAsset.assetId]
+            if (!otherDenomFee) return bn(0)
+            return bnOrZero(otherDenomFee.amountCryptoBaseUnit).times(rate)
+          }
+
+          const firstHopNetworkFee = firstStep.feeData.networkFeeCryptoBaseUnit
+          // Subtract the first-hop network fees for OSMO -> ATOM, which aren't automagically subtracted in the multi-hop abstraction
+          return bnOrZero(firstHopNetworkFee)
+        })()
+      : bn(0)
```

Comment:
Mostly ignore this change, it's just copied from the duplicate `_getReceiveSideAmountsCryptoBaseUnit` function we have elsewhere, that was not updated when the other one was.